app/controllers/providers/edit_initial_allocations_controller.rb

```diff
@@ -0,0 +1,125 @@
+module Providers
+  class EditInitialAllocationsController < ApplicationController
+    def do_you_want
+      if request.post?
+        case params[:allocation][:request_type]
+        when AllocationsView::RequestType::INITIAL
+          redirect_to number_of_places_provider_recruitment_cycle_allocation_path(
+            provider_code: allocation.accredited_body.provider_code,
+            recruitment_cycle_year: recruitment_cycle.year,
+            training_provider_code: allocation.provider.provider_code,
+            id: allocation.id,
+          )
+        when AllocationsView::RequestType::DECLINED
+          destroy
+          redirect_to confirm_deletion_provider_recruitment_cycle_allocation_path(
+            provider_code: provider.provider_code,
+            recruitment_cycle_year: recruitment_cycle.year,
+            training_provider_code: training_provider.provider_code,
+          )
+        end
+      else
+        render locals: {
+          training_provider: training_provider,
+          allocation: allocation,
+          provider: provider,
+        }
+      end
+    end
+
+    def confirm_deletion
+      @allocation = Allocation.new(request_type: AllocationsView::RequestType::DECLINED)
+      @training_provider = training_provider
+      @provider = provider
+      @recruitment_cycle = recruitment_cycle
+      render template: "providers/allocations/show"
+    end
+
+    def number_of_places
+      if params.dig("allocation", "number_of_places")
+        allocation.number_of_places = params.dig("allocation", "number_of_places")
+      end
+
+      if request.post? && allocation.valid?
+        redirect_to check_answers_provider_recruitment_cycle_allocation_path(
+          provider_code: allocation.accredited_body.provider_code,
+          recruitment_cycle_year: recruitment_cycle.year,
+          training_provider_code: allocation.provider.provider_code,
+          number_of_places: params[:allocation][:number_of_places],
+          id: allocation.id,
+        )
+      else
+        render locals: {
+          training_provider: training_provider,
```

Comment:
In the `allocation` method below we include the `provider` in our request to retrieve the allocation, which is the training provider, so could we use that here and anywhere else we need `training_provider` so that we can avoid making an extra request to the API?

I think a good approach could be to use a `build_allocation` hook to get the `allocation`, and then a `build_training_provider` hook to use the provider set in the allocation, if that makes sense.